packages/sqlite/src/index.ts

```diff
@@ -118,115 +92,91 @@ class DatabaseConnectionImplementation implements DatabaseConnection {
 
   async tx<T>(fn: (db: DatabaseTransaction) => Promise<T>): Promise<T> {
     return this._mutex.writeLock(async () => {
-      await new Promise<void>((resolve, reject) => {
-        this._database.run('BEGIN', (err) => {
-          if (err) reject(err);
-          else resolve();
-        });
-      });
+      // TODO extract those as class variables
+      // however I do not know how to get
+      this._begin.run();
       try {
-        const result = fn(
+        const result = await fn(
           new DatabaseTransactionImplementation(this._database),
         );
-        await new Promise<void>((resolve, reject) => {
-          this._database.run('COMMIT', (err) => {
-            if (err) reject(err);
-            else resolve();
-          });
-        });
+        this._commit.run();
         return result;
       } catch (ex) {
-        await new Promise<void>((resolve, reject) => {
-          this._database.run('ROLLBACK', (err) => {
-            if (err) reject(err);
-            else resolve();
-          });
-        });
+        this._rollback.run();
         throw ex;
       }
     });
   }
 
   async dispose() {
-    await new Promise<void>((resolve, reject) => {
-      this._database.close((err) => {
-        if (err) reject(err);
-        else resolve();
-      });
-    });
+    this._database.close();
   }
 }
 
 export default function connect(
   filename: string = IN_MEMORY,
-  options: DatabaseConnectionOptions = {},
+  options: DatabaseOptions = {},
 ): DatabaseConnection {
   return new DatabaseConnectionImplementation(filename, options);
 }
 module.exports = Object.assign(connect, {
   default: connect,
-  DatabaseConnectionMode,
   IN_MEMORY,
   sql,
   isSqlQuery,
 });
 
 async function runQuery(
   query: SQLQuery,
-  database: sqlite.Database,
+  database: Database,
   lock: <T>(fn: () => Promise<T>) => Promise<T>,
 ) {
   const {text, values} = query.format(sqliteFormat);
   return lock(async () => {
-    return await new Promise<any[]>((resolve, reject) => {
-      database.all(text, values, (err, rows) => {
-        if (err) reject(err);
-        else resolve(rows);
-      });
-    });
+    const stm = database.prepare(text);
+    try {
+      return stm.all(...values);
+    } catch (_err) {
+      const err = _err as Error;
+      // TODO we should be able to catch this before calling all()
+      if (err.message.indexOf('This statement does not return data') >= 0) {
+        return stm.run(...values);
+      }
+      throw err
```

Comment:
I do think we'll want to figure out a better way to handle this. I'm not familiar with the `.lzz` language used by better-sqlite3, but it looks like there could be a returns_data property/method to check: https://github.com/WiseLibs/better-sqlite3/blob/6da9e569dbd3ad6d3e04b4cbd61c788c039d2ef3/src/util/query-macros.lzz#L54